src/test/java/org/mastodon/mamut/clustering/ClusterRootNodesControllerTest.java

```diff
@@ -0,0 +1,273 @@
+import org.mastodon.mamut.model.Model;
+import org.mastodon.mamut.model.ModelGraph;
+import org.mastodon.mamut.model.Spot;
+import org.mastodon.mamut.treesimilarity.tree.BranchSpotTree;
+import org.mastodon.model.tag.TagSetStructure;
+import org.mockito.Mockito;
+
+import java.util.Collection;
+import java.util.List;
+
+import static junit.framework.TestCase.assertEquals;
+
+public class ClusterRootNodesControllerTest
+{
+	@Test
+	public void testCreateTagSet()
+	{
+		final Model model = new Model();
+
+		final ModelGraph modelGraph = model.getGraph();
+
+		addLineageTree1( modelGraph );
+		addLineageTree2( modelGraph );
+		addLineageTree3( modelGraph );
+		addLineageTree4( modelGraph );
+		addLineageTree5( modelGraph );
+		addEmptyTree( modelGraph );
+
+		@SuppressWarnings( "unchecked" )
+		ClusterRootNodesListener< BranchSpotTree > listener = Mockito.mock( ClusterRootNodesListener.class );
+
+		ClusterRootNodesController controller = new ClusterRootNodesController( model );
+		controller.addListener( listener );
+		controller.setClusteringMethod( ClusteringMethod.AVERAGE_LINKAGE );
+		controller.setSimilarityMeasure( SimilarityMeasure.NORMALIZED_DIFFERENCE );
+		controller.setCropCriterion( CropCriteria.TIMEPOINT );
+		controller.setCropStart( 0 );
+		controller.setCropEnd( 100 );
+		controller.setMinCellDivisions( 1 );
+		controller.setNumberOfClasses( 3 );
+		controller.createTagSet();
+
+		List< TagSetStructure.TagSet > tagSets = model.getTagSetModel().getTagSetStructure().getTagSets();
+		TagSetStructure.TagSet tagSet0 = model.getTagSetModel().getTagSetStructure().getTagSets().get( 0 );
+		List< TagSetStructure.Tag > tags = tagSet0.getTags();
+		TagSetStructure.Tag tag0 = tags.get( 0 );
+		TagSetStructure.Tag tag1 = tags.get( 1 );
+		TagSetStructure.Tag tag2 = tags.get( 2 );
+
+		Collection< Spot > tag0Spots = model.getTagSetModel().getVertexTags().getTaggedWith( tag0 );
+		Collection< Spot > tag1Spots = model.getTagSetModel().getVertexTags().getTaggedWith( tag1 );
+		Collection< Spot > tag2Spots = model.getTagSetModel().getVertexTags().getTaggedWith( tag2 );
+
+		Mockito.verify( listener, Mockito.times( 1 ) ).clusterRootNodesComputed( Mockito.any(), Mockito.any(), Mockito.anyDouble() );
+		assertEquals( 1, tagSets.size() );
+		assertEquals( 3, tags.size() );
+		assertEquals( "Class 1", tag0.label() );
+		assertEquals( "Class 2", tag1.label() );
+		assertEquals( "Class 3", tag2.label() );
+		assertEquals( 9, tag0Spots.size() );
+		assertEquals( 14, tag1Spots.size() );
+		assertEquals( 12, tag2Spots.size() );
+	}
+
+	/**
+	 * <pre>
+	 *                             branchSpot1(lifespan=20)
+	 *                    ┌-─────────┴─────────────┐
+	 *                    │                        │
+	 *                  branchSpot2(lifespan=10)    branchSpot3(lifespan=30)
+	 * </pre>
+	 */
+	private static void addLineageTree1( final ModelGraph modelGraph )
+	{
+		Spot spot1 = modelGraph.addVertex();
+		spot1.init( 0, new double[ 3 ], 0 );
```

Comment:
I disagree that this is easier to read per se. It might be easier to read, if one is used to it. 

If not (like in my case), I find it confusing that a method with the name "init" returns a reference to the object. I would assume a method called "init" to be void and setting some inner states of the object. Assuming this, it is confusing to hand over the result of an "init" method to an object.

Nevertheless, I made the suggested change here, since it saves lines of code and I was not aware in this case that init() returns a reference to the Object.

cf: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/9a02be6326e2a40832cd5c2586c54101aa6587a8